File: GeeksForGeeks/Trees/Longest-Consecutive-Path-in-Binary-Tree.py
"""
Problem: Longest Consecutive Path in Binary Tree
Platform: GeeksForGeeks
Difficulty: Medium

Approach:
Perform DFS on the tree.
For each node, check if it continues a consecutive sequence
from its parent. Update the maximum length found.

Time Complexity: O(n)
Space Complexity: O(h)
"""

class Solution:
    def longestConsecutive(self, root):
        self.ans = 0

        def dfs(node, parent_val, length):
            if not node:
                return

            if parent_val is not None and node.data == parent_val + 1:
                length += 1
            else:
                length = 1

            self.ans = max(self.ans, length)

            dfs(node.left, node.data, length)
            dfs(node.right, node.data, length)

        dfs(root, None, 0)
        return self.ans

File: GeeksForGeeks/Trees/Longest-Consecutive-Path-in-Binary-Tree.js
/**
 * Problem: Longest Consecutive Path in Binary Tree
 * Platform: GeeksForGeeks
 * Difficulty: Medium
 *
 * Approach:
 * Perform DFS while tracking the parent value and current
 * consecutive path length.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(h)
 */

class Solution {
    longestConsecutive(root) {
        let ans = 0;

        const dfs = (node, parentVal, length) => {
            if (!node) return;

            if (parentVal !== null && node.data === parentVal + 1) {
                length++;
            } else {
                length = 1;
            }

            ans = Math.max(ans, length);

            dfs(node.left, node.data, length);
            dfs(node.right, node.data, length);
        };

        dfs(root, null, 0);

        return ans;
    }
}